dev-tools/validation/test_implementation_capability_tracking.py

```
@@ -0,0 +1,577 @@
+
+def expect_failure(label: str, call, expected: str) -> None:
+    try:
+        call()
+    except SystemExit as error:
+        if expected not in str(error):
+            raise AssertionError(f"{label}: unexpected failure: {error}") from error
+    else:
+        raise AssertionError(f"{label}: invalid fixture unexpectedly passed")
+
+
+def main() -> None:
```

Ruff (0.15.21)

[warning] Too many statements (63 > 50)

(PLR0915)